### PR TITLE
generate file map

### PR DESCRIPTION
**Pull request by OsmondJiang**

### CODE REVIEW

File: src/docfx/build/Build.cs [hunk 1]
```diff
@@ -100,13 +102,12 @@ private static bool ShouldBuildFile(Document itemToBuild, ConcurrentDictionary<D
                 return false;
             }
 
-            if (!set.TryAdd(itemToBuild, 0))
+            if (itemToBuild.ContentType == ContentType.TableOfContents && !tocMap.Contains(itemToBuild))
```
Comment (herohua) [inline src/docfx/build/Build.cs]: >!tocMap.Contains(itemToBuild) [](start = 74, length = 29)

how can this happen? should we assert?

Comment (OsmondJiang) [reply src/docfx/build/Build.cs]: nested toc should not be built, i didn't change the logic here, but switch the order.

---
In reply to: [189461762](https://github.com/dotnet/docfx/pull/2787#discussion_r189461762) [](ancestors = 189461762)

Comment (herohua) [reply src/docfx/build/Build.cs]: I see. looks good to me.

---
In reply to: [189479687](https://github.com/dotnet/docfx/pull/2787#discussion_r189479687) [](ancestors = 189479687,189461762)

File: src/docfx/build/legacy/Legacy.cs [hunk 2]
```diff
@@ -5,28 +5,35 @@
...
+                var fileItem = LegacyFileMapItem.Instance(legacyOutputFilePathRelativeToSiteBasePath, document.ContentType);
+                if (fileItem != null)
+                {
+                    fileMapItems.Add((Path.GetRelativePath(docset.Config.SourceBaseFolder, document.FilePath), fileItem));
+                }
+
+                if (document.ContentType == ContentType.TableOfContents)
                 {
-                    var toc = JsonUtility.Deserialize<LegacyTableOfContentsModel>(File.ReadAllText(file));
+                    var toc = JsonUtility.Deserialize<LegacyTableOfContentsModel>(File.ReadAllText(absoluteOutputFilePath));
```
Comment (yufeih) [inline src/docfx/build/legacy/Legacy.cs]: Extract `ConvertLegacyTableOfContents` to a separate class/method

File: src/docfx/config/Config.cs [hunk 3]
```diff
@@ -40,6 +40,11 @@ internal class Config
         /// </summary>
         public readonly string SiteBasePath = string.Empty;
 
+        /// <summary>
+        /// Just for backward compatibility, the source base folder
+        /// </summary>
+        public readonly string SourceBaseFolder = string.Empty;
```
Comment (yufeih) [inline src/docfx/config/Config.cs]: SourceBasePath?

**Pull request merged**